master/internal/logpattern/logpattern.go

```diff
@@ -76,23 +80,48 @@ func (l *LogPatternPolicies) monitor(ctx context.Context,
 			}
 
 			if compiledRegex.MatchString(log.Log) {
-				switch policy.Action().GetUnionMember().(type) {
-				case expconf.LogActionCancelRetries:
-					if err := addDontRetry(
-						ctx, model.TaskID(log.TaskID), *log.AgentID, policy.Pattern(), log.Log,
-					); err != nil {
-						return fmt.Errorf("adding don't retry: %w", err)
+				if policy.Action() != nil {
+					switch policy.Action().Type {
+					case expconf.LogActionTypeCancelRetries:
+						if err := addDontRetry(
+							ctx, model.TaskID(log.TaskID), *log.AgentID, *pattern, log.Log,
+						); err != nil {
+							return fmt.Errorf("adding don't retry: %w", err)
+						}
+
+					case expconf.LogActionTypeExcludeNode:
+						if err := addRetryOnDifferentNode(
+							ctx, model.TaskID(log.TaskID), *log.AgentID, *pattern, log.Log,
+						); err != nil {
+							return fmt.Errorf("adding retry on different node: %w", err)
+						}
+					default:
+						return fmt.Errorf("unrecognized log pattern policy type")
 					}
+				}
 
-				case expconf.LogActionExcludeNode:
-					if err := addRetryOnDifferentNode(
-						ctx, model.TaskID(log.TaskID), *log.AgentID, policy.Pattern(), log.Log,
-					); err != nil {
-						return fmt.Errorf("adding retry on different node: %w", err)
+				if policy.Name() != nil {
+					err = db.Bun().RunInTx(ctx, nil, func(ctx context.Context, tx bun.Tx) error {
```

Comment:
ideally this wouldn't be in a separate transaction

Reply:
Maybe i don’t understand the ideal approach you mentioned yet. My intention is to put two update queries in the same transaction, so they either both succeed or failed

Reply:
sorry, i was totally unclear. i agree these two query should be in a transaction, but i think there are also writes in `addRetryOnDifferentNode` and `addDontRetry` that are in a separate transaction.

Reply:
I see. Queries of one policy are either all succeed or all fail.